bindings/pyroot_experimental/PyROOT/src/RTensorPyz.cxx

```diff
@@ -0,0 +1,167 @@
+// Author: Stefan Wunsch CERN  07/2019
+// Original PyROOT code by Wim Lavrijsen, LBL
+
+/*************************************************************************
+ * Copyright (C) 1995-2018, Rene Brun and Fons Rademakers.               *
+ * All rights reserved.                                                  *
+ *                                                                       *
+ * For the licensing terms see $ROOTSYS/LICENSE.                         *
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#include "CPyCppyy.h"
+#include "CPPInstance.h"
+#include "ProxyWrappers.h"
+#include "PyROOTPythonize.h"
+#include "RConfig.h"
+#include "TInterpreter.h"
+
+#include <sstream>
+
+////////////////////////////////////////////////////////////////////////////
+/// \brief Adopt memory of a Python object with array interface using an RTensor
+/// \param[in] obj PyObject with array interface
+///
+/// This function returns an RTensor which adopts the memory of the given
+/// PyObject. The RTensor takes the data pointer and the shape from the array
+/// interface dictionary.
+PyObject *PyROOT::AsRTensor(PyObject * /*self*/, PyObject * obj)
```

Comment:
I guess you saw this question coming :smile:  but this function has quite some code repetition with respect to `AsRVec`. Could you please check how to outsource the common parts to a helper method?

Reply:
I'm gonna give it a try ;)

Reply:
I don't want to move the error handling out of the function since it complicates more than it helps. However, I've put in a shared function for the conversion of the numpy type string to the respective C++ type. All other parts actually look similar logic-wise but share little information, e.g., the code string we have to jit at the end.

Reply:
What is exactly the complication? Can't you have a helper function that indicates you if there was some error, and if so you return null? Seriously, there is a lot of duplicated parts...

Reply:
Factored out common operations between `AsRTensor` and `AsRVec`!